services/venice-server/src/main/java/com/linkedin/venice/grpc/GrpcRequestContext.java

```diff
@@ -0,0 +1,93 @@
+package com.linkedin.venice.grpc;
+
+import com.linkedin.davinci.listener.response.ReadResponse;
+import com.linkedin.venice.listener.ServerStatsContext;
+import com.linkedin.venice.listener.request.RouterRequest;
+import com.linkedin.venice.response.VeniceReadResponseStatus;
+import io.grpc.stub.StreamObserver;
+
+
+/**
+ * We need to keep track of each request as it goes through the pipeline so that we can record the necessary metrics
+ * and separate different parts of the logic for the response. If a specific handler raises an error, we set
+ * the hasError flag to true and stop executing the rest of the pipeline excluding the stats collection.
+ */
+public class GrpcRequestContext<T> {
+  private final StreamObserver<T> responseObserver;
+  private final Class<?> responseType;
```

Comment:
Can we use an enum instead? I think it would be cleaner when inspecting this property to be able to do a switch/case on the enum rather than a series of `if (responseType instanceof xyz)`. And if we do need the `Class` object, we can still get it out of a property of the enum.

Reply:
I agree, and originally it was written that way, but I switched to using responseType to avoid introducing yet another enum. However, that’s not a strong enough reason. So switching back to enum

Reply:
Done